internal/webhooks/clusterclass.go

```diff
@@ -123,8 +142,90 @@ func (webhook *ClusterClass) validate(old, new *clusterv1.ClusterClass) error {
+	allErrs = append(allErrs, webhook.validateRemovedMachineDeploymentClassesAreNotUsed(old, new)...)
+
 	if len(allErrs) > 0 {
 		return apierrors.NewInvalid(clusterv1.GroupVersion.WithKind("ClusterClass").GroupKind(), new.Name, allErrs)
 	}
 	return nil
 }
+
+func (webhook *ClusterClass) validateRemovedMachineDeploymentClassesAreNotUsed(old, new *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+	if old == nil {
+		return allErrs
+	}
+	removedClasses := webhook.removedMachineClasses(old, new)
+
+	// If no classes have been removed return early as no further checks are needed.
+	if len(removedClasses) == 0 {
+		return allErrs
+	}
+
+	// Retrieve all clusters using the ClusterClass.
+	clusters, errs := webhook.getClustersUsingClusterClass(old)
+	if len(errs) > 0 {
+		allErrs = append(allErrs, errs...)
+		return allErrs
+	}
+
+	// Create a set of records of machineDeploymentClass information where the name is a fully qualified name for the
+	// MachineDeployment topology in the form cluster.Name/machineDeploymentTopology.Name and links to the machineDeploymentClass Name.
+	for _, c := range clusters {
+		for _, machineDeploymentTopology := range c.Spec.Topology.Workers.MachineDeployments {
+			if removedClasses.Has(machineDeploymentTopology.Class) {
+				// TODO(killianmuldoon): Improve error printing here so large scale changes don't flood the error log e.g. deduplication, only example usages given.
+				allErrs = append(allErrs, field.Forbidden(field.NewPath(""), fmt.Sprintf("MachineDeploymentClass %v is in use in MachineDeployment %v in Cluster %v. ClusterClass %v modification not allowed",
+					machineDeploymentTopology.Class, machineDeploymentTopology.Name, c.Name, old.Name),
+				))
+			}
+		}
+	}
+
+	return allErrs
+}
+
+func (webhook *ClusterClass) removedMachineClasses(old, new *clusterv1.ClusterClass) sets.String {
+	removedClasses := sets.NewString()
+
+	// Ensure no MachineDeploymentClass in use has been removed.
+	classes := webhook.classNamesFromWorkerClass(new.Spec.Workers)
+	for _, oldClass := range old.Spec.Workers.MachineDeployments {
+		if !classes.Has(oldClass.Class) {
+			removedClasses.Insert(oldClass.Class)
+		}
+	}
+	return removedClasses
+}
+
+// classNames returns the set of MachineDeployment class names.
+func (webhook *ClusterClass) classNamesFromWorkerClass(w clusterv1.WorkersClass) sets.String {
+	classes := sets.NewString()
+	for _, class := range w.MachineDeployments {
+		classes.Insert(class.Class)
+	}
+	return classes
+}
+
+func (webhook *ClusterClass) getClustersUsingClusterClass(clusterClass *clusterv1.ClusterClass) ([]clusterv1.Cluster, field.ErrorList) {
+	var allErrs field.ErrorList
+	clusters := &clusterv1.ClusterList{}
+	clustersUsingClusterClass := []clusterv1.Cluster{}
+	err := webhook.Client.List(context.Background(), clusters,
+		client.MatchingLabels{
+			clusterv1.ClusterTopologyOwnedLabel: "",
+		},
+		client.InNamespace(clusterClass.Namespace),
```

Comment:
You can use a field selector with the index `ClusterClassNameField` to get all clusters of a ClusterClass.

The downside is that the field selector is not unit-testable with the fake client, but I would still use it.

Reply:
This is a sticky one to implement as I'm going to have to rewrite the unit tests - I should have an update in a little while.

Reply:
As discussed, we deferred that one

Reply:
#5721 This has been pushed until the issue to enable envtest with webhooks has been resolved.